PowerShell Binding

FuzzySecurity opened the pull request:
Adds the PowerShell binding to Keystone. Feel free to edit as required to match the repo style!

aquynh commented:
this is awesome, but a question: how was Invoke-Keystone.ps1 created? is there anyway to generate that binary blob inside with some script, so whenever the core is changed we can regenerate it again ourself?

FuzzySecurity commented:
Hola, yes there is a script which can be used to generate the compressed DLL's. I'll add it to the pull request!

aquynh commented:
great, now what do you think about removing the binary file Invoke-Keystone.ps1, and let users to create that when they first use this binding?

FuzzySecurity commented:
Mm that is a bit tricky because the script only generates compressed output for the Emit-Keystone function and you need to manually update (twice, for x32/64). I could rewrite the script to generate the entire Emit-Keystone function but then I think the binding will also need to be updated to:

- Add a check to see if "Emit-Keystone" exist, else tell the user to generate it.
- Remove the version flag/ASCII banner as it would not be possible to say, a priori, what version the binding uses.

Let me know what you think, I can make the change if you like!

aquynh commented:
i am not PS user, but in principal, my idea is simple: we avoid having binary blob, and let users generate that themselves. this is because the code on Github is always changing, so the blob is always outdated.

however, we can provide the binary blob when we release a new version, so the PS binding is readily available for users. they can always regenerate the blob if they dont trust provided binary.

mattifestation commented:
Personally, for a binding like this, I would advocate for providing the source for the C# portion of the binding for the sake of transparency. I'd like to be able to read the source and learn more about the implementation without necessarily having to dump the embedded binary and decompiling it. While I can certainly sympathize with wanting to contain everything in an embedded package for offensive purposes, are there scenarios you envision where keeping this purely in memory would be ideal over loading the C# DLL portion from disk?

FuzzySecurity commented:
@aquynh agreed, I'll rework the script so users need to first generate the binary blob.

@mattifestation the c# is inline in the script and can be reviewed (also not very big). The binary blobs are a compressed version of the pre-compiled keystone DLL's here:
http://www.keystone-engine.org/download/

Let me put together a new pull request for review!

mattifestation commented:
Ah okay. Apologies for misunderstanding. Great work, BTW!

aquynh commented:
as suggested by @mattifestation, what do you think about loading the DLL at runtime, so we dont need to embed it?

mattifestation commented:
@FuzzySecurity For reference, I was calling SetDllDirectory accordingly based on the architecture of the running process. https://github.com/mattifestation/capstone/blob/master/bindings/csharp/Capstone/capstone.cs#L163

FuzzySecurity commented:
After some consideration, I decided to remove the function which writes the Keystone DLL to disk altogether. Basically the user needs to compile the keystone DLL or grab the precompiled version from the project website. After the DLL is placed in a SafeDllSearchMode directory the binding can be used as normal.

- Updated the README
- Reworked binding a bit
- Added try/catch to detect if the DLL is missing (sample screenshot)

![trycatch](https://cloud.githubusercontent.com/assets/10997340/20026979/5dfd84ba-a307-11e6-9d9d-04d19b22332a.png)

aquynh commented:
excellent! now just one minor thing: what do you think about generating the constants KS_xxx automatically from source, rather than maintaining yourself? the reason is that i add new stuff to the core frequently, and want to avoid bothering the binding authors as much as possible (so ease the burden of binding maintainers)

this is what i have been doing: whenever there is a minor change in `include/keystone/*.h`, i do this and all the bindings are updated, automatically:

```
$ cd bindings
$ make
```

the magic lies in file `bindings/const_generator.py`, which is called for each binding, and auto-gen the const files.

now if you can add PS support to const_generator.py, that will be awesome. 
or i can help you on that.

aquynh commented:
the first step towards this is to separate the const symbols from Invoke-Keystone.ps1, and put them into other files. i suppose that is simple enough?

FuzzySecurity commented:
I'm not sure how to go about that to be honest, also I'm using a mix between inline c# and PowerShell. The binding only uses/needs a select few constants from "keystone.h", ks_err ([here](https://github.com/FuzzySecurity/keystone/blob/2d8bee32f970d2af823b1da001885e1970545756/bindings/powershell/Invoke-Keystone.ps1#L104)), ks_arch ([here](https://github.com/FuzzySecurity/keystone/blob/2d8bee32f970d2af823b1da001885e1970545756/bindings/powershell/Invoke-Keystone.ps1#L196)), ks_mode ([here](https://github.com/FuzzySecurity/keystone/blob/2d8bee32f970d2af823b1da001885e1970545756/bindings/powershell/Invoke-Keystone.ps1#L210)), ks_opt_value ([here](https://github.com/FuzzySecurity/keystone/blob/2d8bee32f970d2af823b1da001885e1970545756/bindings/powershell/Invoke-Keystone.ps1#L232)). ks_err would also need to be updated in place, separating it out would break the runtime c# compilation.

Not really sure how to proceed, maybe @mattifestation has an idea? I don't really mind updating it from time to time, it would only take a minute or two.

mattifestation commented:
Yeah it would be cool to update const_generator.py to support C# and PowerShell in order to avoid some of the costs of maintenance. My opinion is that a PowerShell binding for a project like this ideally should be in the form of a module. While it hasn't been updated you may want to check out the layout of my Unicorn binding (https://github.com/mattifestation/UnicornPowerShell). I modeled the layout after the other bindings and the consts files kind of lend themselves to being updated with something like const_generator.py (assuming someone updates it).

FuzzySecurity commented:
@mattifestation has anyone told you that you have terrible bedside manner? :p

Ok so, briefly:
  * Renamed to Get-KeystoneAssembly
  * Converted into a PowerShell Module
  * DLL's are, again, contained in the binding
  * The constants have been separated out into individual files "Keystone\Const\*.cs"

@aquynh Can you have a look if the format can be integrated with const_generator.py

mattifestation commented:
@FuzzySecurity Alright then. I'll be sure to provide a "compliment sandwich" for all subsequent code reviews. :P

FuzzySecurity commented:
@mattifestation jk, well deserved, I'm admittedly a pretty bad developer ;)

@aquynh should be all working now! Can you check my addition to "const_generator.py", the format I need is:
NAME = INT,

I'm only grabbing "keystone.h" as I don't need any other constant.

aquynh commented:
it looks great now, but what do you think about removing the DLLs totally, and ask users to download, or compile themselves instead? the reason is like before: DLLs are always outdated.

FuzzySecurity commented:
@aquynh that should do it

FuzzySecurity commented on bindings/powershell/Keystone/Keystone.psm1:
```diff
@@ -124,9 +124,16 @@ function Get-KeystoneAssembly {
+		$DllPath = $($PSScriptRoot + '\Lib\Keystone\x86\keystone.dll').Replace('\','\\')
 	} else {
-		$DllPath = $($PSScriptRoot + '\Lib\Keystone-0.9.1\x64\keystone.dll').Replace('\','\\')
+		$DllPath = $($PSScriptRoot + '\Lib\Keystone\x64\keystone.dll').Replace('\','\\')
```
Ok, I changed it. The script would only have checked the DLL in use anyway.

aquynh commented:
looks great to me! @mattifestation, do you have any more comments?

mattifestation commented:
@FuzzySecurity 

**Looks pretty good! The changes look great!**
Could you consider putting -Version in its own parameter set. That parameter is mutually exclusive from the rest of the parameters
**Again, thanks for incorporating all these changes.**

See what I did there? :P

FuzzySecurity commented:
@mattifestation haha indeed indeed :P, not sure if I missed something though but "-Version" is already in it's own parameter set?

mattifestation commented:
Yes sir. You are correct. Perhaps I was looking at an older version. Good stuff!!!

aquynh commented:
looks awesome! i just have some minor comments

- will you remove that .keep file?
- please add PS suppor to bindings/Makefile, so `cd bindings; make` would auto-gen const files of PS. (this is the behaviour of all other bindings now)

FuzzySecurity commented:
@aquynh I just added the ".keep" file because otherwise that folder is not added to the pull request (because it is empty till the user adds the DLL). As people tend to make mistakes, I think it is practical that the empty folder is already present in the download.

As for the Makefile, I'm not sure how to add my entry, is it just like this?
```
all:
	cd python && $(MAKE) gen_const
	cd nodejs && $(MAKE) gen_const
	python const_generator.py rust
	python const_generator.py go
	cd ruby && $(MAKE) gen_const
	python const_generator.py powershell
```

aquynh commented:
I think you can rename that .keep to .gitignore for the same purpose.

Looks like this change to Makefile is right. Will test it later.

FuzzySecurity commented:
Cool, I made the changes, yea please check the constant generating is ok as I'm not too familiar with that part and I don't currently have access to a build environment.

aquynh commented:
i made a minor fix for const_generator.py here: https://github.com/keystone-engine/keystone/commit/c1386ba2c1daf3c772f14999cc6362d8bf4b177b

merged, thanks for your amazing effort: https://github.com/keystone-engine/keystone/commit/76652911bd0d7d6e3a5a660cba0360464464cd21!

would you make the same change for the Capstone binding, too?

FuzzySecurity commented:
Awesome, thanks for putting up with the many changes :D! Yes I had the same idea, I'm already converting the Capstone module for integration.
